Remove default arguments before calling to __torch_dispatch__

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #61123

This applies the design pattern of removing explicit arguments when they
coincide with the default arguments.  This simplifies argument patterns
that dispatch kernels receive and make it easier for us to maintain BC
(as addition of a new default argument isn't immediately BC-breaking
for dispatch implementors).

There is an important extra API which I haven't implemented here yet,
which is to take an incomplete sequence of arguments and fill out their
defaults (in case the user did want normalization).  I plan on adding
that in a future PR.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D29853616](https://our.internmc.facebook.com/intern/diff/D29853616)

ezyang committed:
Remove default arguments before calling to __torch_dispatch__

This applies the design pattern of removing explicit arguments when they
coincide with the default arguments.  This simplifies argument patterns
that dispatch kernels receive and make it easier for us to maintain BC
(as addition of a new default argument isn't immediately BC-breaking
for dispatch implementors).

There is an important extra API which I haven't implemented here yet,
which is to take an incomplete sequence of arguments and fill out their
defaults (in case the user did want normalization).  I plan on adding
that in a future PR.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

bdhirsh reviewed (approved):
LGTM!

bhosmer reviewed (approved):
LG, though I'm curious about hidden eq test expense, per inline

ezyang commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -1498,17 +1498,34 @@ void concrete_dispatch_fn(const c10::impl::PyInterpreter*, const c10::OperatorHa
+    }
+    const auto& default_ivalue = *arg.default_value();
+    const auto& ivalue = arguments[idx];
+    if (default_ivalue != ivalue) {
```
I don't think any of our ivalues are interned so most of the time you aren't going to manage to hit pointer equality.

While the equality test could be a large structural test, defaults are always in practice relatively small structs (e.g., small list). An appropriately written equality test should be able to bail out early in such cases in O(1).

bhosmer replied:
Well, the worry would be the cases that *are* equal, so no early bailout on inequality (unless you mean something else). But you're right, if there's no interning then there's no point in pointer checking. I pictured whoever was filling in the defaults upstream taking direct references to `arg.default_value()` values, but of course mutable params.

And yeah our limited defaulting syntax does cap the risk for now - like, you can't specify a tensor default. Future enhancements to schema might cause stealthy perf regressions here but I guess we'll have to cross that bridge.

ezyang replied:
If something is equal, it is equal *with the default value*, and so you're still doing O(1) because default values are small.

albanD reviewed (approved):
LGTM

I wonder if it is important to mention that the new args must be added at the end of the schema for this to work?

Also is there a plan to create a nice BC detection test suite for `__torch_dispatch__`? I feel like it would be relatively easy to do to ensure that we don't break third party libs?

ezyang commented:
> I wonder if it is important to mention that the new args must be added at the end of the schema for this to work?

The fact that we cannot add new keyword-only args in any order without BC breaking is a problem (and also see https://github.com/pytorch/pytorch/issues/61664 ). In this particular case I think this can be fixed but I have to think carefully about the design.

ezyang committed:
Update on "Remove default arguments before calling to __torch_dispatch__"


This applies the design pattern of removing explicit arguments when they
coincide with the default arguments.  This simplifies argument patterns
that dispatch kernels receive and make it easier for us to maintain BC
(as addition of a new default argument isn't immediately BC-breaking
for dispatch implementors).

There is an important extra API which I haven't implemented here yet,
which is to take an incomplete sequence of arguments and fill out their
defaults (in case the user did want normalization).  I plan on adding
that in a future PR.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29853616).
